source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -43,57 +54,116 @@ template <class RequestType, class ResponseType> class GrpcMuxFailover {
       GrpcStreamCallbacks<ResponseType>* stream_callbacks)>;
 
   GrpcMuxFailover(GrpcStreamCreator primary_stream_creator,
-                  OptRef<GrpcStreamCreator> failover_stream_creator,
+                  absl::optional<GrpcStreamCreator> failover_stream_creator,
                   GrpcStreamCallbacks<ResponseType>& grpc_mux_callbacks)
       : grpc_mux_callbacks_(grpc_mux_callbacks), primary_callbacks_(*this),
-        primary_grpc_stream_(std::move(primary_stream_creator(&primary_callbacks_))) {
+        primary_grpc_stream_(std::move(primary_stream_creator(&primary_callbacks_))),
+        connecting_to_primary_(false), connected_to_primary_(false), connecting_to_failover_(false),
```

Comment:
Should these state variables be a single enum (at least the mutually exclusive ones)?

Reply:
In the near future the only 2 mutually exclusive ones are going to be connected_to_primary and connected_to_failover. I'm not sure if making the "connected" an enum and the "connecting" boolean flags makes the code more readable.
[Previous comment](https://github.com/envoyproxy/envoy/pull/31383#discussion_r1434205551) when a single non-exclusive flags were used, suggested using independent bools.

Reply:
My vote is for `connected_to` to be enum and the others boolean. Rather than readability, having structural guards against bugs here seems a net good in itself.

Reply:
Changed to an enum for connected_to and ever_connected_to.